parser/src/sourcemanager.cpp

```diff
@@ -221,6 +234,30 @@ void SourceManager::updateFile(const model::File& file_)
     });
 }
 
+void SourceManager::removeFile(const model::File& file_)
+{
+  bool removeContent = false;
+
+  // Delete File and FileContent (only when no other File references it)
+  _transaction([&, this]() {
+    auto relFiles = _db->query<model::File>(
+        odb::query<model::File>::content == file_.content.object_id());
+    if(std::distance(relFiles.begin(), relFiles.end()) == 1)
+    {
+      removeContent = true;
+      _db->erase<model::FileContent>(file_.content.object_id());
+    }
+    _db->erase<model::File>(file_.id);
+  });
+
+  // Maintain cache
+  std::lock_guard<std::mutex> guard(_createFileMutex);
+  _files.erase(file_.path);
```

Comment:
For visual indication of the critical section, because it's not the whole function which should be *synchronized*, please put the statements into a new block:

```cpp
  std::lock_guard<std::mutex> guard(_createFileMutex);
  {
    _files.erase(file_.path);
    // ...
  }
}

Reply:
Everything should be synchronized after the creation of the guard, and that is the visual indication. (That is 4 lines of code, it's not like one can miss it visually.)
Also, the visual indication of a critical section should match it functionally, so in that case the guard should be created inside the new block I think.

Reply:
Yeah, do that. Must have had some Java on my mind with the whole `synchronized`.

Reply:
Fixed in 64cb407581662e78675baf916a95dba2b765b3e4.